chrysom/client.go

```diff
@@ -23,19 +23,45 @@ import (
 	"encoding/json"
 	"errors"
 	"fmt"
+	"io"
 	"io/ioutil"
 	"net/http"
+	"strings"
 	"time"
 
 	"github.com/go-kit/kit/log"
 	"github.com/go-kit/kit/log/level"
 	"github.com/go-kit/kit/metrics/provider"
-	"github.com/xmidt-org/argus/model"
 	"github.com/xmidt-org/argus/store"
 	"github.com/xmidt-org/bascule/acquire"
 	"github.com/xmidt-org/themis/xlog"
 )
 
+const storeAPIPath = "/api/v1/store"
+
+// Errors that can be returned by this package. Since some of these errors are returned wrapped, it
+// is safest to use errors.Is() to check for them.
+var (
+	ErrAddressEmpty             = errors.New("argus address is required")
+	ErrBucketEmpty              = errors.New("bucket name is required")
+	ErrItemIDEmpty              = errors.New("item ID is required")
+	ErrItemIDMismatch           = errors.New("item ID must match that in payload")
+	ErrItemDataEmpty            = errors.New("data field in item is required")
+	ErrUndefinedMetricsProvider = errors.New("a metrics provider is required")
+	ErrUndefinedIntervalTicker  = errors.New("interval ticker is nil. Can't listen for updates")
+	ErrGetItemsFailure          = errors.New("failed to get items. Non-200 statuscode was received")
+	ErrRemoveItemFailure        = errors.New("failed to delete item. Non-200 statuscode was received")
+	ErrPushItemFailure          = errors.New("failed to push item. Non-success statuscode was received")
```

Comment:
I thought about making these more specific but they would blow up in number so I thought they'd still be helpful if returned in some form like this:
```
return nil, fmt.Errorf("statusCode %v: %w", response.Code, ErrPushItemFailure)
```
This will help when looking at tr1d1um logs and realizing Argus is returning a 403 instead of just a generic non-200 which I think is what you were getting at.

Reply:
I have no problem with error variables; my point is all of them are exported.  The only reason to export an error is to provide consumers of this package a way to do different logic based on receiving the error:
```
err := doSomething()
// special flow
if errors.Is(err, ErrNotFound) {
    doSomethingElse()
    return
}
// normal flow
```